src/main/java/io/xdb/core/exception/XDBHttpException.java

```diff
@@ -0,0 +1,47 @@
+package io.xdb.core.exception;
+
+import feign.FeignException;
+import io.xdb.core.encoder.ObjectEncoder;
+import io.xdb.gen.models.ApiErrorResponse;
+import io.xdb.gen.models.EncodedObject;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+public abstract class XDBHttpException extends RuntimeException {
+
+    private final int statusCode;
+    private ApiErrorResponse apiErrorResponse;
+
+    public XDBHttpException(final ObjectEncoder objectEncoder, final FeignException.FeignClientException exception) {
+        super(exception);
+        statusCode = exception.status();
+
+        String decodeErrorMessage = "";
+        final Optional<ByteBuffer> respBody = exception.responseBody();
+        if (respBody.isPresent()) {
+            final String data = StandardCharsets.UTF_8.decode(respBody.get()).toString();
+            try {
+                apiErrorResponse = objectEncoder.decode(new EncodedObject().data(data), ApiErrorResponse.class);
+                return;
+            } catch (final Exception e) {
+                decodeErrorMessage = e.getMessage();
+            }
+        }
+
+        // TODO
+        apiErrorResponse =
+            new ApiErrorResponse().detail("empty or unable to decode to apiErrorResponse: " + decodeErrorMessage);
+    }
+
+    public static XDBHttpException fromFeignException(
+        final ObjectEncoder objectEncoder,
+        final FeignException.FeignClientException exception
+    ) {
+        if (exception.status() >= 400 && exception.status() < 500) {
+            return new ClientSideException(objectEncoder, exception);
```

Comment:
for 400<statusCode<500, can you add some more friendly exceptions (use final exceptions):
404: ProcessNotFoundException
409: ProcessAlreadyStartedException 

This is also something that I want to improve in iWF: https://github.com/indeedeng/iwf-java-sdk/issues/203

and we have done the same in Python SDK later: https://github.com/indeedeng/iwf-python-sdk/blob/main/iwf/errors.py